### ISSUE TITLE
Edgetpu compilation problem for custom detection model

### ISSUE THREAD

**Issue by ostrowskimarcin** [tensorflow/models#10695]
Hi,
I downloaded some pretrained models from TF1 ZOO such as spaghettinet, mobiledet, etc. When I compile their included .tflite versions with edgetpu_compiler everything is fine, almost every operation is mapped properly. After training models with custom dataset and converting it to tflite, many layers aren't being mapped in comparison to included model. Problem occures also with different model types. Here's example of compilation log:
### 1. pretrained model (spaghettinet_l)
```

Edge TPU Compiler version 16.0.384591198
Input: model.tflite
Output: model_edgetpu.tflite

Operator                       Count      Status

STRIDED_SLICE                  10         Mapped to Edge TPU
CONCATENATION                  7          Mapped to Edge TPU
MAX_POOL_2D                    3          Mapped to Edge TPU
RESHAPE                        11         Mapped to Edge TPU
ADD                            31         Mapped to Edge TPU
CUSTOM                         1          Operation is working on an unsupported data type
LOGISTIC                       1          Mapped to Edge TPU
CONV_2D                        107        Mapped to Edge TPU
SLICE                          19         Mapped to Edge TPU
RESIZE_NEAREST_NEIGHBOR        3          Mapped to Edge TPU
DEPTHWISE_CONV_2D              43         Mapped to Edge TPU

```
### 2. custom model (spaghettinet_l)
```
Edge TPU Compiler version 16.0.384591198
Input: model.tflite
Output: model_edgetpu.tflite

Operator                       Count      Status

DEPTHWISE_CONV_2D              43         Mapped to Edge TPU
SLICE                          19         Mapped to Edge TPU
LOGISTIC                       1          Mapped to Edge TPU
RESHAPE                        11         Mapped to Edge TPU
CUSTOM                         1          Operation is working on an unsupported data type
STRIDED_SLICE                  10         Mapped to Edge TPU
CONCATENATION                  7          Mapped to Edge TPU
RESIZE_NEAREST_NEIGHBOR        3          Mapped to Edge TPU
DEQUANTIZE                     12         Operation is working on an unsupported data type
CONV_2D                        102        Mapped to Edge TPU
CONV_2D                        5          Operation is working on an unsupported data type
QUANTIZE                       8          Mapped to Edge TPU
QUANTIZE                       5          Operation is otherwise supported, but not mapped due to some unspecified limitation
MAX_POOL_2D                    3          Mapped to Edge TPU
ADD                            5          Operation is working on an unsupported data type
ADD                            26         Mapped to Edge TPU

```
I made some basic changes in pipeline.config, such as number of classes, batch size, number of steps, etc. I doubt those changes had an impact on operations and compilation but here's how it looks like:

```
# SpaghettiNet Feature Extractor optimized for EdgeTPU.
# Trained on COCO17 from scratch.
#
# spaghettinet_edgetpu_l
# Achieves 28.02% mAP on COCO17 at 400k steps.
# 1.75ms EdgeTPU latency at 1.57 billion MACs.
#
# TPU-compatible.

model {
  ssd {
    inplace_batchnorm_update: true
    freeze_batchnorm: false
    num_classes: 1
    box_coder {
      faster_rcnn_box_coder {
        y_scale: 10.0
        x_scale: 10.0
        height_scale: 5.0
        width_scale: 5.0
      }
    }
    matcher {
      argmax_matcher {
        matched_threshold: 0.5
        unmatched_threshold: 0.5
        ignore_thresholds: false
        negatives_lower_than_unmatched: true
        force_match_for_each_row: true
        use_matmul_gather: true
      }
    }
    similarity_calculator {
      iou_similarity {
      }
    }
    encode_background_as_zeros: true
    anchor_generator {
      ssd_anchor_generator {
        num_layers: 5
        min_scale: 0.2
        max_scale: 0.95
        aspect_ratios: 1.0
        aspect_ratios: 2.0
        aspect_ratios: 0.5
        aspect_ratios: 3.0
        aspect_ratios: 0.3333333
      }
    }
    image_resizer {
      fixed_shape_resizer {
        height: 160
        width: 160
      }
    }
    box_predictor {
      convolutional_box_predictor {
        min_depth: 0
        max_depth: 0
        num_layers_before_predictor: 0
        use_dropout: false
        dropout_keep_probability: 0.8
        kernel_size: 3
        use_depthwise: true
        box_code_size: 4
        apply_sigmoid_to_scores: false
        class_prediction_bias_init: -4.6
        conv_hyperparams {
          activation: RELU_6,
          regularizer {
            l2_regularizer {
              weight: 0.00004
            }
          }
          initializer {
            random_normal_initializer {
              stddev: 0.03
              mean: 0.0
            }
          }
          batch_norm {
            train: true,
            scale: true,
            center: true,
            decay: 0.97,
            epsilon: 0.001,
          }
        }
      }
    }
    feature_extractor {
      type: 'ssd_spaghettinet'
      spaghettinet_arch_name: 'spaghettinet_edgetpu_l'
      use_explicit_padding: false
    }
    loss {
      classification_loss {
        weighted_sigmoid_focal {
          alpha: 0.75,
          gamma: 2.0
        }
      }
      localization_loss {
        weighted_smooth_l1 {
          delta: 1.0
        }
      }
      classification_weight: 0.5
      localization_weight: 1.5
    }
    normalize_loss_by_num_matches: true
    normalize_loc_loss_by_codesize: true
    post_processing {
      batch_non_max_suppression {
        score_threshold: 1e-8
        iou_threshold: 0.6
        max_detections_per_class: 10
        max_total_detections: 10
        use_static_shapes: true
      }
      score_converter: SIGMOID
    }
  }
}

train_config: {
  batch_size: 4
  sync_replicas: true
  startup_delay_steps: 0
  replicas_to_aggregate: 32
  num_steps: 22447
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
  data_augmentation_options {
    random_vertical_flip {
    }
  }
  optimizer {
    momentum_optimizer: {
      learning_rate: {
        cosine_decay_learning_rate {
          learning_rate_base: 0.8
          total_steps: 22447
          warmup_learning_rate: 0.13333
          warmup_steps: 200
        }
      }
      momentum_optimizer_value: 0.9
    }
    use_moving_average: false
  }
  max_number_of_boxes: 10
  unpad_groundtruth_tensors: false
}

train_input_reader: {
  label_map_path: "label_map.pbtxt"
  shuffle: true
  tf_record_input_reader {
    input_path: "dataset/train/train.record-?????-of-00010"
  }
}

eval_config: {
  metrics_set: "coco_detection_metrics"
  use_moving_averages: false
  num_examples: 5010
}

eval_input_reader: {
  label_map_path: "label_map.pbtxt"
  shuffle: true
  num_epochs: 1
  tf_record_input_reader {
    input_path: "dataset/val/val.record-?????-of-00010"
  }
}

graph_rewriter {
  quantization {
    delay: 40000
    weight_bits: 8
    activation_bits: 8
  }
}

```
After training I create tflite model with following code:

```bash
#!/bin/bash

# Exit script on error.
set -e
# Echo each command, easier for debugging.
set -x

# -------------CONSTANTS---------------- #

MAIN_DIR="$PWD/test_models/spaghettinet"

CKPT_NUMBER="10000"
SIZE=160

INPUT_TENSORS='normalized_input_image_tensor'
OUTPUT_TENSORS='TFLite_Detection_PostProcess,TFLite_Detection_PostProcess:1,TFLite_Detection_PostProcess:2,TFLite_Detection_PostProcess:3'

TRAIN_DIR="${MAIN_DIR}/train"
OUTPUT_DIR="${MAIN_DIR}/tflite_models"

# ------------------------------------- #

mkdir -p "${OUTPUT_DIR}"

echo "EXPORTING frozen graph from checkpoint..."
python3 models/research/object_detection/export_tflite_ssd_graph.py \
  --pipeline_config_path="${MAIN_DIR}/pipeline.config" \
  --trained_checkpoint_prefix="${TRAIN_DIR}/model.ckpt-${CKPT_NUMBER}" \
  --output_directory="${OUTPUT_DIR}" \
  --add_postprocessing_op=true

echo "CONVERTING frozen graph to TF Lite file..."
tflite_convert \
  --output_file="${OUTPUT_DIR}/model.tflite" \
  --graph_def_file="${OUTPUT_DIR}/tflite_graph.pb" \
  --inference_type=QUANTIZED_UINT8 \
  --input_arrays="${INPUT_TENSORS}" \
  --output_arrays="${OUTPUT_TENSORS}" \
  --mean_values=128 \
  --std_dev_values=128 \
  --input_shapes=1,$SIZE,$SIZE,3 \
  --allow_custom_ops \
  --change_concat_input_ranges=false \
  --allow_nudging_weights_to_use_fast_gemm_kernel=true \
  --enable_v1_converter

echo "TFLite graph generated at ${OUTPUT_DIR}/model.tflite"

```

I'm using TF 1.15 installed with nvidia  [repo](https://github.com/NVIDIA/tensorflow), but I face the issue also with Colab. I checked compilation for original model size (320) but it's not the point.

Does anybody have a similar issue? I'm wondering maybe it's not because of pipeline.config, but one of converting functions or their arguments? Thanks in advance.

**Comment by muxamilian** [comment]
Try this Jupyter Notebook: https://gist.github.com/NobuoTsukamoto/eade17835e57a02f5414aae907293707
I could get it to work with the normal tensorflow 1.15, not nvidia's. If you want to use Spaghettinet you have to patch tensorflow and the object detection API quite a lot (all the patches are included in the notebook). 

Also useful: 

- https://github.com/tensorflow/tensorflow/pull/55919
- https://discuss.tensorflow.org/t/training-a-spaghettinet-model/8648
- https://www.reddit.com/r/tensorflow/comments/u1c6in/why_are_google_releasing_new_models_based_on_tf1/

**Comment by ostrowskimarcin** [comment]
@muxamilian thank you, it helped a lot! Unfortunately tf 1.15 is no longer supported on Google Colab, so I couldn't test it there. Your second link helped me to solve the issue. Exporting spaghettinet graph to tflite also requires following flags:
  --default_ranges_min=0 \
  --default_ranges_max=6 \
This worked for me.

Now edgetpu compilation output looks like this:
```
Edge TPU Compiler version 16.0.384591198
Input: model.tflite
Output: model_edgetpu.tflite

Operator                       Count      Status

CONCATENATION                  7          Mapped to Edge TPU
RESIZE_NEAREST_NEIGHBOR        3          Mapped to Edge TPU
CONV_2D                        107        Mapped to Edge TPU
DEPTHWISE_CONV_2D              43         Mapped to Edge TPU
QUANTIZE                       8          Mapped to Edge TPU
MAX_POOL_2D                    3          Mapped to Edge TPU
ADD                            31         Mapped to Edge TPU
LOGISTIC                       1          Mapped to Edge TPU
STRIDED_SLICE                  10         Mapped to Edge TPU
RESHAPE                        11         Mapped to Edge TPU
SLICE                          19         Mapped to Edge TPU
DEQUANTIZE                     2          Operation is working on an unsupported data type
CUSTOM                         1          Operation is working on an unsupported data type
```
Except CUSTOM operation (which cannot be mapped even in pretrained model) there are only 2 DEQUANTIZE ops. I assume it's okay, because of floating point output, but correct me if I'm wrong. Maybe there are other things to change

**Comment by ostrowskimarcin** [comment]
Except CUSTOM operation (which cannot be mapped even in pretrained model) there are only 2 DEQUANTIZE ops. I assume it's okay, because of floating point output, but correct me if I'm wrong. Maybe there are other things to change?

**Comment by muxamilian** [comment]
@ostrowskimarcin Btw, I tested SpaghettiNet and MobileDet and even though SpaghettiNet is supposed to perform better, in reality it doesn't. Accuracy is always worse than for MobileDet. Furthermore, MobileDet also doesn't require any patching etc. So I would go for that.
